go/libkb/secret_store_file.go

```diff
@@ -37,44 +70,138 @@ func (s *SecretStoreFile) RetrieveSecret(username NormalizedUsername) (LKSecFull
 	return newLKSecFullSecretFromBytes(secret)
 }
 
+func (s *SecretStoreFile) retrieveSecretV2(username NormalizedUsername) (LKSecFullSecret, error) {
+	xor, err := ioutil.ReadFile(s.userpathV2(username))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return LKSecFullSecret{}, ErrSecretForUserNotFound
+		}
+
+		return LKSecFullSecret{}, err
+	}
+
+	noise, err := ioutil.ReadFile(s.noisepathV2(username))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return LKSecFullSecret{}, ErrSecretForUserNotFound
+		}
+
+		return LKSecFullSecret{}, err
+	}
+
+	var xorFixed secretBytes
+	copy(xorFixed[:], xor)
+	var noiseFixed noiseBytes
+	copy(noiseFixed[:], noise)
+	secret, err := s.noiseXOR(xorFixed, noiseFixed)
+	if err != nil {
+		return LKSecFullSecret{}, err
+	}
+
+	return newLKSecFullSecretFromBytes(secret)
+}
+
+func (s *SecretStoreFile) makeNoise() (noiseBytes, error) {
+	var nb noiseBytes
+	noise, err := RandBytes(noiseLen)
+	if err != nil {
+		return nb, err
+	}
+	copy(nb[:], noise)
+	return nb, nil
+}
+
+func (s *SecretStoreFile) noiseXOR(secret secretBytes, noise noiseBytes) ([]byte, error) {
+	sum := sha256.Sum256(noise[:])
+	if len(sum) != len(secret) {
+		return nil, errors.New("LKSecLen or sha256.Size is no longer 32")
+	}
+
+	xor := make([]byte, len(sum))
```

Comment:
Yes, that's the idea.  We want to be able to "shred" the secret even on SSDs, which of course make it hard due to FTLs.  So the idea is that the attacker would have to recreate the 2mb "noise" file bit-for-bit to recover the secret, rather than just finding the unlinked physical block. It still may be doable, but slightly harder than the naive approach.